File: Week3/2 - Valid Parenthesis String/Solution.py
#
# Created on Thu Apr 16 2020
#
# Title: Leetcode - Valid Parenthesis String
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

class Solution:
    def checkValidString(self, s):
        # Check if string is empty, if empty string then return True as empty string is given valid
        if s == "":
            return True
        
        # Create 2 list representing stacks.
        # main_stack for pushing index of left parenthesis.
        # star_stack for pushing index of star elements.
        main_stack = list()
        star_stack = list()

        # Iterate over string
        for index, val in enumerate(s):
            if val == '(':
                main_stack.append(index)
            elif val == '*':
                star_stack.append(index)
            
            # else case triggered whenever ')' is found. 
            else:

                # pop from main_stack if not empty
                if not len(main_stack) == 0:
                    main_stack.pop()
                
                # pop from star_stack if not empty. here * works like '(' 
                elif not len(star_stack) == 0:
                    star_stack.pop()
                
                # else case means no more * left to convert to '(', hence return False
                else:
                    return False
        
        # balancing the '(' in main_stack
        while not len(main_stack) == 0:
            if len(star_stack) == 0:
                return False
            elif main_stack[-1] < star_stack[-1]:
                main_stack.pop()
                star_stack.pop()
            else:
                return False
        
        return True

# Sample inputs.
obj = Solution()
print(obj.checkValidString("(*))"))
print(obj.checkValidString("(*)"))
print(obj.checkValidString("())"))

File: Week3/2 - Valid Parenthesis String/Solution.cpp
/*
 * Created on Thu Apr 16 2020
 *
 * Title: Leetcode - Valid Parenthesis String
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

#include <iostream>
#include <stack>

using namespace std;

class Solution {
public:
    bool checkValidString(string s) {
        int n = s.size();

        if(n == 0) {
            return true;
        }    
        stack<int> main_stack, star_stack;

        for(int i=0; i<n; ++i) {
            if(s[i] == '(') {
                main_stack.push(i);
            }
            else if(s[i] == '*') {
                star_stack.push(i);
            }
            else {
                if(!main_stack.empty()) {
                    main_stack.pop();
                }
                else if(!star_stack.empty()) {
                    star_stack.pop();
                }
                else {
                    return false;
                }
            }
        }

        while(!main_stack.empty()) {
            if(star_stack.empty()) {
                return false;
            }
            else if(main_stack.top() < star_stack.top()) {
                main_stack.pop();
                star_stack.pop();
            }
            else {
                return false;
            }
        }
        return true;    
    }
};